crates/rust-analyzer/src/diagnostics/to_proto.rs

```diff
@@ -84,6 +87,30 @@ fn diagnostic_related_information(
     Some(lsp_types::DiagnosticRelatedInformation { location, message })
 }
 
+/// Resolves paths mimicking VSCode's behavior when `file_name` starts
+/// with the root directory component, which does not discard the base
+/// path.  If this relative path exists, use it, otherwise fall back
+/// to the existing Rust behavior of path joining.
+fn resolve_path(workspace_root: &Path, file_name: &str) -> PathBuf {
+    let file_name = Path::new(file_name);
+
+    // Test path with VSCode's path join behavior.
+    let vscode_path = {
+        let mut result = PathBuf::from(workspace_root);
+        result.extend(file_name.components().skip_while(|component| match component {
+            std::path::Component::RootDir => true,
+            _ => false,
+        }));
+        result
+    };
+    if vscode_path.exists() {
```

Comment:
This thing does a syscall, so it's slow and not-pure. That's generally not what we want (although for diagnostics this might be fine).

Is there some way to predict from just looking at the path which behavior do we want here?

Reply:
I wasn't sure if this would be allowed but figured I'd throw it up and see how you responded 😉

Specific to my use case I could just check for leading `//` but I don't think that's really an established convention.  Buck/Bazel use that for root relative to a label, so I'm just piggybacking off that in how I'm referring to workspace absolute paths.  (See https://docs.bazel.build/versions/master/build-ref.html#labels for an example.)

Seeing as you're matching the URL against a mapping you've already built up, is there a way I could do the equivalent `exists()` check against the mapping instead of with a file system call?